managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/PendingReadsManager.java

```diff
@@ -0,0 +1,450 @@
+                return new PendingReadKey(other.endEntry + 1, endEntry);
+            }
+            return null;
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static final class ReadEntriesCallbackWithContext {
+        final AsyncCallbacks.ReadEntriesCallback callback;
+        final Object ctx;
+        final long startEntry;
+        final long endEntry;
+    }
+
+    @AllArgsConstructor
+    private static final class FindPendingReadOutcome {
+        final PendingRead pendingRead;
+        final PendingReadKey missingOnLeft;
+        final PendingReadKey missingOnRight;
+        boolean needsAdditionalReads() {
+            return missingOnLeft != null || missingOnRight != null;
+        }
+    }
+
+    private FindPendingReadOutcome findPendingRead(PendingReadKey key, Map<PendingReadKey,
+            PendingRead> ledgerCache, AtomicBoolean created) {
+        synchronized (ledgerCache) {
+            PendingRead existing = ledgerCache.get(key);
+            if (existing != null) {
+                COUNT_PENDING_READS_MATCHED.inc(key.size());
+                COUNT_ENTRIES_NOTREAD_FROM_BK.inc(key.size());
+                return new FindPendingReadOutcome(existing, null, null);
+            }
+            FindPendingReadOutcome foundButMissingSomethingOnLeft = null;
+            FindPendingReadOutcome foundButMissingSomethingOnRight = null;
+            FindPendingReadOutcome foundButMissingSomethingOnBoth = null;
+
+            for (Map.Entry<PendingReadKey, PendingRead> entry : ledgerCache.entrySet()) {
+                PendingReadKey entryKey = entry.getKey();
+                if (entryKey.includes(key)) {
+                    COUNT_PENDING_READS_MATCHED_INCLUDED.inc(key.size());
+                    COUNT_ENTRIES_NOTREAD_FROM_BK.inc(key.size());
+                    return new FindPendingReadOutcome(entry.getValue(), null, null);
+                }
+                if (entryKey.overlaps(key)) {
+                    PendingReadKey reminderOnLeft = key.reminderOnLeft(entryKey);
+                    PendingReadKey reminderOnRight = key.reminderOnRight(entryKey);
+                    if (reminderOnLeft != null && reminderOnRight != null) {
+                        foundButMissingSomethingOnBoth = new FindPendingReadOutcome(entry.getValue(),
+                                reminderOnLeft, reminderOnRight);
+                    } else if (reminderOnRight != null && reminderOnLeft == null) {
+                        foundButMissingSomethingOnRight = new FindPendingReadOutcome(entry.getValue(),
+                                null, reminderOnRight);
+                        // we can exit the loop in this case, as below we are not going to
+                        // consider the other options
+                        break;
+                    } else if (reminderOnLeft != null && reminderOnRight == null) {
+                        foundButMissingSomethingOnLeft = new FindPendingReadOutcome(entry.getValue(),
```

Comment:
Sorry, I don't get why we don't break here? Any reason `right` is better than `left`?
From what I see, there are some aspects to consider :
1. Less `rangeEntryCache.asyncReadEntry0` calls is better. So `left` and `right` > `both`.
2. Less cycles in this for loop. So just break if we find any matched result.
3. Less entries to read from BK. Which we are not considering here. This needs more cycles or other complex data structures.

Reply:
below we have to choose what happens in case we have:
- foundButMissingSomethingOnLeft
- foundButMissingSomethingOnRight
- foundButMissingSomethingOnBoth

Currently we give preference to foundButMissingSomethingOnRight. So we can break the loop when we found it.

We can add in the future something more sophisticated to decide "the best" operation to reuse.

I did a few tests and  "missing on right" is the thing that happens more frequently, both on tailing reads and when I have many subscriptions that are catching up (even if I give preference to the other options).
This makes sense as the subscriptions are moving forward so when a new read comes and there is already one pending it is common that the new read misses the new entries "on the right" (on the tail of the topic).

After re-reading the code If we break here when we find "anything" we could lose the chance to find the "match include" case, that is the absolutely the best, as we don't need further reads.
So I am going to remove this "break"

Reply:
> So I am going to remove this "break"


If so, since we are going through all the `PendingReadKey` in ledgerCache. I seems we can simple achieve the goal of "Less entries to read from BK" by comparing the number of entries to read. This can also solve the case that there are multiply `foundButMissingSomethingOnLeft` or `foundButMissingSomethingOnRight`.